servo/connectors/newrelic.py

```diff
@@ -0,0 +1,362 @@
+    #             response = await client.get(newrelic_request.url)
+    #             response.raise_for_status()
+    #             result = response.json()
+    #             return f"returned {len(result)} results"
+
+    #     return self.config.metrics, query_for_metric
+
+@metadata(
+    description="Newrelic Connector for Opsani",
+    version="0.0.1",
+    homepage="https://github.com/opsani/newrelic-connector",
+    license=License.APACHE2,
+    maturity=Maturity.EXPERIMENTAL,
+)
+class NewrelicConnector(BaseConnector):
+    """NewrelicConnector objects enable servo assemblies to capture
+    measurements from the [Newrelic](https://newrelic.com/) metrics server.
+    """
+    config: NewrelicConfiguration
+
+    # TODO
+    # @on_event()
+    # async def check(self,
+    #     filter_: Optional[Filter] = None, 
+    #     halt_on: HaltOnFailed = HaltOnFailed.requirement
+    # ) -> List[Check]:
+    #     """Checks that the configuration is valid and the connector can capture        
+    #     measurements from Newrelic.
+
+    #     Checks are implemented in the NewrelicChecks class.
+
+    #     Args:
+    #         filter_ (Optional[Filter], optional): A filter for limiting the
+    #             checks that are run. Defaults to None.
+    #         halt_on (HaltOnFailed, optional): When to halt running checks.
+    #             Defaults to HaltOnFailed.requirement.
+
+    #     Returns:
+    #         List[Check]: A list of check objects that report the outcomes of the            
+    #             checks that were run.
+    #     """        
+    #     return await NewrelicChecks.run(self.config, filter_, halt_on=halt_on)
+
+    @on_event()
+    def describe(self) -> Description:
+        """Describes the current state of Metrics measured by querying Newrelic.
+
+        Returns:
+            Description: An object describing the current state of metrics
+                queried from Newrelic.
+        """
+        return Description(metrics=self.config.metrics)
+
+    @property
+    @on_event()
+    def metrics(self) -> List[Metric]:
+        """Returns the list of Metrics measured through Newrelic queries.
+
+        Returns:
+            List[Metric]: The list of metrics to be queried.
+        """
+        return self.config.metrics
+
+    @on_event()
+    def instances(self) -> List[str]:
+        async with self.api_client as client:
+            try:
+                response = await client.get('/config')
+                response.raise_for_status()
+            except (httpx.HTTPError, httpcore._exceptions.ReadTimeout, httpcore._exceptions.ConnectError) as error:                
+                self.logger.trace(f"HTTP error encountered during oco config GET {self.optimizer.api_url}/config: {error}")
+                raise
+
+        opsani_config = response.json()
+        canary = opsani_config.get('adjustment', {}).get('control', {}).get('userdata', {}).get('canary')
```

Comment:
Note that with oco-e, we are using a new "Assets API" to store/retrieve component data with a new url and data structure:

`url = "https://api.opsani.com/accounts/{}/applications/{}/assets/opsani.com/<COMPONENT>".format(
    config["opsani_account"], config["opsani_app_name"]
)`

`canary = opsani_config.get('data')
`

Reply:
@benburdick what does COMPONENT refer to in this case? I could have sworn that was tied to the adjustment and I'm not sure its easily accessed from the measure connector under the current model

Reply:
@linkous8 Component in this case would be "canary" - in the servo-harness connector its referenced as `comp_key`

Reply:
We are going to need to revisit the `comp_name`/`component` stuff anyway. The current naming convention precludes us from being able to do multi-connector (i.e. the artist formerly known as aggregation) setups. I think that we may be able to factor out the reliance on the assets endpoint entirely which would be a big win because the data is non-authoritative but this discussion is trending more toward Harness than core New Relic concerns.

I would recommend that to start wading into these waters we simplify the implementation in the core New Relic connector by returning data for all instances unless another connector responds to the `instances` event, getting the rest of the implementation of this port landed, and then bringing up a Harness connector on a new branch that starts with nothing but an `instances` event handler.

We can start with the existing implementation using the IP address hostname computation stuff and then evolve it once we validate some of the changes @benburdick and I have been discussing